### ISSUE TITLE
Filter against CTI types

### ISSUE THREAD

**Issue by brudil** [joist-orm/joist-orm#722]
Not sure how easy this would be - I've seen the type assigned within the query Joist produces but haven't investigated any more than that.

You can 'filter' via providing a sub type:
```ts
em.find(SmallPublisher, { where })
```

But if you're wanting to filter against the parent/root, you have no way to limit by the type. Conditionals within the Entity param can break generic types.

Something like: 
```ts
em.find(Publisher, { type: SmallPublisher })
```

Decisions to be consider on what is passed in and the name of the filter property (as there might be a userland field of that name)

**Comment by stephenh** [comment]
Hi @brudil ; yep, I haven't done a ton of thinking around this, other than that the current setup works but should definitely be fancier/more robust.

I.e. right, to limit by type, you have to just know to use the `SmallPublisher` as the first `find` arg, which is not very ergonomic.

We do have precedent for sneaking in extra `find` filter keys and :crossed_fingers: that they don't overlap with userland attributes, see the `as` key for binding aliases.

I mean, it's probably "too cute" but we could do like `t: ...`, with the idea that `t` (or `st` for "subtype") is so short surely no one has a userland key of that :upside_down_face: :-) 

I dunno, maybe we should establish a convention of prefixes for Joist-special filter keys, like `$as` and `$type`, but I've really enjoyed avoiding those so far...

As you said, `type` is the most natural but also seems very likely to conflict with a userland `type` field/enum/etc., also as you mentioned...

> Conditionals within the Entity param can break generic types.

I don't doubt this, but was not 100% following; do you have an example?

The other limitation I don't have a great solution for, but would like to support, is doing `em.find(Publisher, ...)` and then being able to use fields from subtypes, like being able to query `SmallPublisher`s or `LargePublisher`s in the same query. Obviously you couldn't have an `AND` with fields from both subtypes, as that would never match a real row, but I'm not sure if we really need to enforce that in the `em.find` type system, vs. it just being kind of implicit/obvious hopefully.

But, right, I briefly thought about trying to make `em.find(Publisher, FilterOf<Publisher | LargePublisher | SmallPublisher)` work but didn't get that far, and haven't personally needed it yet...

All said, if we dial this issue into just filtering by sub-type...and I had to implement it today...I would probably go with `t` or `st` as a special keyword and just doc that this is an opinionated/ergonomic convention of Joist, similar to `as`...

Wdyt?

**Comment by brudil** [comment]
> I don't doubt this, but was not 100% following; do you have an example?

Here's an example from our codebase;
<img width="1034" alt="image" src="https://github.com/stephenh/joist-ts/assets/382352/0bcf3b79-fa32-4ed4-8ac9-4e1436f5babe">

Although explictly passing in the generic does work as a union, so scratch that really.


We have a Filter <-> GraphQL exposure layer to handle "computed" filters, like "expired" etc. Which I've rewritten to produce complex conditions to be passed to Joist's where. So I'm thinking for the time being I'll be able to handle the basic filtering on types by allowing that system to provide the sub type given to Joist.

Allowing this is be considered a little more. The multi-type querying sounds helpful, but we're thankfully not needing that either just yet.

**Comment by brudil** [comment]
Actually, I'm wondering if the current types a bit optimistic or there is a bug.

Where you might have two sub types and the parent has say, an `organisation` relation. Performing something like 

```
em.find(ASubType, { organisation: SomeOrganisationEntity })
```
This is valid because the sub type filter interface extends the parent types.

but it looks like somewhere along the line, it's not picked up as a parent type properly, resulting in the query:

```sql
select "tsp".*, tsp_b0.*, "tsp".id as id from "subtype" as "tsp" left outer join "parenttype" as "tsp_b0" on "tsp"."id" = "tsp_b0"."id" where "tsp"."organisation_id" = $1 order by "tsp"."id" ASC limit $2 - column tsp.organisation_id does not exist
```

**Comment by stephenh** [comment]
Ah yep...that `tsp`-should-be-`tsp_b0` should be fixed by #723.

**Comment by stephenh** [comment]
We've a version of this now via #1093 where you can filter to subtype-specific fields like:

* `em.find(Author, { publisherLargePublisher: { country: "US" } })`

Where the pattern is:

* `publisher` / `relation` name +
* `LargePublisher` / `subtype` name

It's a little long, but the combination of `relation` first + `subtype` last is guaranteed to be unique and not clash with other relations, so going with that for now.

@brudil I know this isn't quite what you were asking for, insofar as you're wanting to do an `em.find(SmallPublisher or LargePublisher, where)` and have the `where` typed appropriately on the result of the 1st condition.

I feel like this is probably possible with sufficient TS magic, but my guess/hope is that most of it would just be in some sort of caller/semi-cast-ish incantation and not changing the `em.find` directly.

So going to close this out for now--the current PR only supported m2os, and we do need to get to the other relations.

@brudil given you're an active user of CTI, feel free to file issues to apply the similar `publisher(Large|SmallPublisher)` trick for other non-m2o relation types, and that will bump them up in the priority list :-), otherwise we'll implement them as needed.

Thanks!
